### PR TITLE
Add `Cosign.ClaimedIdentity` API

### PR DESCRIPTION
**Pull request by saschagrunert**
#### Summary

We now feature a `payload.Cosign.ClaimedIdentity` field which can be used to customize the `image.Critical.Identity.DockerReference`.

Refers to https://github.com/sigstore/cosign/pull/2984, https://github.com/containers/image/issues/1952
#### Release Note

* Added `payload.Cosign.ClaimedIdentity` field which can be used to customize the `image.Critical.Identity.DockerReference`.
#### Documentation

None

### CODE REVIEW

**Review by mtrmac — COMMENTED** [review]
I strongly support the idea of separating the signed image “identity” (what the end-to-end verification claims the image to be) from the image location and digest.

This is similar to how a CA signs “a public key”, but not actually _just_ a public key: the certificate also contains a subject, and that‘s what allows a single trusted CA’s public key to sign public keys of many different subjects, without a risk of confusion.

Here, an image signature signs “a manifest (digest)”, but it should also contain an image identity, and that’s what allows a single trusted signer’s public key to sign many different container images, without a risk of confusion.

---

Compare WIP https://github.com/sigstore/sigstore/compare/main...mtrmac:sigstore:signed-identity , which is a bit more draconian WRT unused code in this area. (I didn’t get to finish the cosign counterpart I’m afraid, so treat that as a tentative WIP that shouldn’t block this PR from proceeding.)

**Review by cpanato — APPROVED** [review]
lgtm

File: pkg/signature/payload/payload.go
Comment (haydentherapper) [reply pkg/signature/payload/payload.go:77]: A breaking change in the api is fine, we just need to document it in the changelog.

**Comment by saschagrunert** [comment]
Updated the PR for testing and API. PTAL again.

**Comment by saschagrunert** [comment]
@cpanato the integration test looks like a flake, may I ask you to retrigger it?

**Comment by haydentherapper** [comment]
Cc @znewman01

File: pkg/signature/payload/payload_test.go
Comment (mtrmac) [inline pkg/signature/payload/payload_test.go:82]: Non-blocking:
```suggestion
				Identity: "docker.io/library/test:1.2.3",
```
and likewise please use tags in the other ones, to be a bit more plausible as examples (fully-qualified, and with tag values) — especially now that these are the first/only actual uses in this repo.

File: pkg/signature/payload/payload.go
Comment (mtrmac) [inline pkg/signature/payload/payload.go:80]: Why is this not located in `SimpleContainerImage`, and specific to `MarshalJSON`?

Comment (saschagrunert) [reply pkg/signature/payload/payload.go]: Moved the whole logic into `SimpleContainerImage` as suggested. :+1:

**Review by cpanato — APPROVED** [review]
lgtm

**Comment by cpanato** [comment]
@haydentherapper do you have any plans to release other things for cosign or we should be good to run a patch release?

File: pkg/signature/payload/payload.go
Comment (mtrmac) [reply pkg/signature/payload/payload.go:56]: https://github.com/sigstore/sigstore/compare/main...mtrmac:sigstore:signed-identity uses `ClaimedIdentity` (along with an extensive comment).

`ClaimedImageIdentity`? `ClaimedImageName`?

Comment (haydentherapper) [reply pkg/signature/payload/payload.go:56]: I like mentioning Image, supportive of either!

Comment (saschagrunert) [reply pkg/signature/payload/payload.go]: @mtrmac I copied over your comment if you don't mind and changed it to become `ClaimedIdentity`.

**Comment by saschagrunert** [comment]
@cpanato @haydentherapper @znewman01 are we happy to move forward with this PR?

**Review by cpanato — APPROVED** [review]
LGTM

thank you

**Comment by znewman01** [comment]
I think you need to mark @haydentherapper's comment as "resolved" (which it is).

**Comment by znewman01** [comment]
CC @bobcallaway for the sigstore/sigstore approvals (Hayden, Carlos, and I have all LGTM'd)

**Review by lukehinds — APPROVED** [review]
lgtm

**Comment by saschagrunert** [comment]
Thank you!

**Pull request merged**
